### PR TITLE
thread_pool: work: notify evenetfd once when idle

### PR DESCRIPTION
**Pull request by bhalevy**
notifying multiple times with no preemption is wasteful, adding extraneous system calls to this path.

Instead, write to _notify_eventfd only once
while _main_thread_idle remains true.

### CODE REVIEW

**Comment by bhalevy** [comment]
Ref #1507

**Comment by xemul** [comment]
It was explicitly made this way (ecf98a507d88af8ea74716806f94f35906e85899). Old code notified the main thread once -- after all syscalls completed.

I think this change makes things worse then they were before the mentioned fix.

1. reactor submits syscall_A and waits
2. reactor submits syscall_B and waits
3. reactor goes to sleep
4. thread executes syscall_A and wakes up reactor
5. reactor resolves syscall_A's future, _but not B's_
6. reactor goes to sleep
7. thread executes syscall_B, _but doesn't wake up reactor_
8. reactor "never" wakes up

**Comment by bhalevy** [comment]
> It was explicitly made this way ([ecf98a5](https://github.com/scylladb/seastar/commit/ecf98a507d88af8ea74716806f94f35906e85899)). Old code notified the main thread once -- after all syscalls completed.
> 
> I think this change makes things worse then they were before the mentioned fix.
> 
> 1. reactor submits syscall_A and waits
> 2. reactor submits syscall_B and waits
> 3. reactor goes to sleep
> 4. thread executes syscall_A and wakes up reactor
> 5. reactor resolves syscall_A's future, _but not B's_
> 6. reactor goes to sleep
> 7. thread executes syscall_B, _but doesn't wake up reactor_

Hmm, you mean the thread will miss the reactor status transitions from idle to non idle and back again, right?

That's a good point.  Could probably be solved with an atomic counter incremented when the reacor wakes up, or a sample of a monotonic clock.

> 8. reactor "never" wakes up

**Comment by avikivity** [comment]
> It was explicitly made this way ([ecf98a5](https://github.com/scylladb/seastar/commit/ecf98a507d88af8ea74716806f94f35906e85899)). Old code notified the main thread once -- after all syscalls completed.
> 
> I think this change makes things worse then they were before the mentioned fix.
> 
> 1. reactor submits syscall_A and waits
> 2. reactor submits syscall_B and waits
> 3. reactor goes to sleep
> 4. thread executes syscall_A and wakes up reactor
> 5. reactor resolves syscall_A's future, _but not B's_
> 6. reactor goes to sleep
> 7. thread executes syscall_B, _but doesn't wake up reactor_
> 8. reactor "never" wakes up

How could this happen? If the reactor went to sleep in 6, we'd wake it up in 7. We check on every wakeup attempt.

**Comment by bhalevy** [comment]
> > > It was explicitly made this way ([ecf98a5](https://github.com/scylladb/seastar/commit/ecf98a507d88af8ea74716806f94f35906e85899)). Old code notified the main thread once -- after all syscalls completed.
> > > I think this change makes things worse then they were before the mentioned fix.
> > > 
> > > 1. reactor submits syscall_A and waits
> > > 2. reactor submits syscall_B and waits
> > > 3. reactor goes to sleep
> > > 4. thread executes syscall_A and wakes up reactor
> > > 5. reactor resolves syscall_A's future, _but not B's_
> > > 6. reactor goes to sleep
> > > 7. thread executes syscall_B, _but doesn't wake up reactor_
> > > 8. reactor "never" wakes up
> > 
> > 
> > How could this happen? If the reactor went to sleep in 6, we'd wake it up in 7. We check on every wakeup attempt.
> 
> Oh, you mean post @bhalevy's patch? I think the patch is just wrong.

I agree that the initial version is wrong, as you both noted.

> We should wake the reactor up after every syscall, since syscalls can take a long time (can involve disk seeks and/or flushes). Waking the reactor up allows it to push more work.

Please look into the latest version I pushed.
It uses an atomic counter to prevent extraneous wake ups if the main thread remains idle during `thread_pool::work` inner loop processing of the completion queue.

File: src/core/thread_pool.hh [hunk 1]
```diff
@@ -52,11 +52,17 @@ public:
...
+        auto new_val = std::abs(_main_thread_idle.load(std::memory_order_relaxed)) + 1;
+        _main_thread_idle.store(new_val, std::memory_order_seq_cst);
+    }
     // When we exit interrupt mode, however, we can safely used relaxed order. If any reordering
     // takes place, we'll get an extra signal and complete will be called one extra time, which is
     // harmless.
-    void exit_interrupt_mode() { _main_thread_idle.store(false, std::memory_order_relaxed); }
+    void exit_interrupt_mode() {
+        auto new_val = -std::abs(_main_thread_idle.load(std::memory_order_relaxed));
+        _main_thread_idle.store(new_val, std::memory_order_relaxed);
```
Comment (avikivity) [inline src/core/thread_pool.hh:64]: Why store -1 here? We know we're existing interrupt mode and the main thread is not idle any more.

Comment (avikivity) [reply src/core/thread_pool.hh:64]: The next time we call enter_interrupt_mode, it will bump the counter to 1.

File: src/core/thread_pool.cc [hunk 2]
```diff
@@ -54,7 +55,9 @@ void thread_pool::work(sstring name) {
 
             // Prevent the following load of _main_thread_idle to be hoisted before the writes to _completed above.
             std::atomic_thread_fence(std::memory_order_seq_cst);
-            if (_main_thread_idle.load(std::memory_order_relaxed) >= 0) {
+            int64_t idle = _main_thread_idle.load(std::memory_order_relaxed);
+            if (idle >= 0 && idle != last_idle) {
+                last_idle = idle;
```
Comment (avikivity) [inline src/core/thread_pool.cc:60]: I don't think it's worth the effort. What are the odds that the syscall completes so quickly the reactor isn't woken up? Typically syscalls require I/O and so sleep.  Moreover, typically the reactor isn't idle (if it is, we can affort an extra syscall).

In any case syscalls are moving to io_uring.

Comment (bhalevy) [reply src/core/thread_pool.cc:60]: okay, makes sense.
I'll retract this PR then

Comment (bhalevy) [reply src/core/thread_pool.cc:60]: I forgot process() invokes the system call via the work_item func.
What I had in mind is that this loop just processes results and therefore does very little.

**Comment by bhalevy** [comment]
retracted

**Pull request closed without merging**